769 replace filters argument

chlebowa opened the pull request:
Fixes #769 

Argument is renamed as per request.
Downstream effects are resolved.

I also identified instances of partial name matching when indexing lists and removed them. There might be more. This is an argument against using `list[["name"]]` over `list$name`. The occurrence of `modules$filter` was an unwelcome surprise.

Also fixes #821

gogonzo commented on R/module_nested_tabs.R:
```diff
@@ -330,11 +327,8 @@ srv_nested_tabs.teal_module <- function(id, datasets, modules, is_module_specifi
+  if (is.null(names(datanames))) {
+    datanames <- structure(datanames, names = datanames)
+  }
+  lapply(datanames, function(x) rlang::hash(datasets$get_data(x, filtered = FALSE)))
```
```suggestion
  lapply(datanames, function(x) rlang::hash(datasets$get_data(x, filtered = FALSE)))
```

Dropping names from the list will cause dropping hashes from the code. See here
https://github.com/insightsengineering/teal/blob/720a374771b296460f0eee9dabf00d4891f7c831/R/get_rcode_utils.R#L73

gogonzo commented on R/module_nested_tabs.R:
```diff
@@ -172,18 +172,16 @@ srv_nested_tabs.teal_modules <- function(id, datasets, modules, is_module_specif
-        )
-      }
-    )
+    modules_reactive <- lapply(names(modules$children), function(module_id) {
+      srv_nested_tabs(
+        id = module_id,
+        datasets = datasets[[labels[module_id]]],
+        modules = modules$children[[module_id]],
+        is_module_specific = is_module_specific,
+        reporter = reporter
+      )
+    })
+    names(modules_reactive) <- names(modules$children)
```
What is wrong with `sapply` in this case? `sapply` adds names for free. In my opinion the only controversy with `sapply` could be `simplify = TRUE` (default), but we can just use `simplify = FALSE`.

gogonzo commented on R/module_nested_tabs.R:
```diff
@@ -292,14 +291,12 @@ srv_nested_tabs.teal_module <- function(id, datasets, modules, is_module_specifi
-    X = datanames,
-    FUN = function(x) eventReactive(trigger_data(), datasets$get_data(x, filtered = TRUE))
-  )
+  data <- lapply(datanames, function(x) eventReactive(trigger_data(), datasets$get_data(x, filtered = TRUE)))
```
`data` needs to be a named list.
```suggestion
  data <- sapply(datanames, simplify = FALSE, function(x) eventReactive(trigger_data(), datasets$get_data(x, filtered = TRUE)))
```

gogonzo commented on R/init.R:
```diff
@@ -141,13 +141,13 @@ init <- function(data,
   resolve_modules_datanames <- function(modules) {
     if (inherits(modules, "teal_modules")) {
-      modules$children <- sapply(modules$children, resolve_modules_datanames, simplify = FALSE)
+      modules$children <- lapply(modules$children, resolve_modules_datanames)
```
Names are dropped here, is this okey? Elements of `teal_modules` should be named

chlebowa commented:
Common answer to `sapply` comments:

> What is wrong with `sapply` in this case? `sapply` adds names for free. In my opinion the only controversy with `sapply` could be `simplify = TRUE` (default), but we can just use `simplify = FALSE`.

> Names are dropped here, is this okey? Elements of `teal_modules` should be named

I change `sapply` to `lapply` as a force of habit, mostly because of simplification results being not entirely secure. It is also frowned upon in some circles, e.g. on Bioconductor.

As you mentioned, `lapply` can reuse the names of `X`, even though it cannot **build** names from the values of `X`. That is why I added names to the character vector that goes into `lapply` whenever possible and where not possible, I added names afterwards. Indeed, the names are even verified by unit tests.

Perhaps I did go a little overboard. If we agree that in all these cases the result should be a list (which is apparently certain but one can never bee too careful), I will restore `sapply` with `simplify = FALSE`.

We could use `Map` -  it calls `mapply` under the hood so names are created like with `sapply` but it always goes `simplify = FALSE`. I say could but I do not necessarily think we should as it is not a commonly used function. We do throw `Filter` around a lot, though.

gogonzo commented:
@chlebowa I think  `sapply(X, simplify = FALSE)` is entirely predictable and useful - we can use single call instead of two when we want to `USE.NAMES`

chlebowa commented:
blocked pending [fixing python app](https://github.com/insightsengineering/teal.gallery/issues/39)

chlebowa commented:
unblocked

gogonzo reviewed (approved):
👍
